handlers/challenge_handler.py

```diff
@@ -603,24 +603,27 @@ class SolveCommand(Command):
     @classmethod
     def execute(cls, slack_wrapper, args, timestamp, channel_id, user_id, user_is_admin):
         """Execute the Solve command."""
-        challenge = ""
-
-        if args:
-            challenge = get_challenge_from_args(ChallengeHandler.DB, args, channel_id)
 
-            if not challenge:
-                challenge = get_challenge_by_channel_id(ChallengeHandler.DB, channel_id)
+        # Get the challenge from the channel id or the first argument to !solve.
+        challenge = get_challenge_from_args(ChallengeHandler.DB, args, channel_id)
+        if challenge:
+            if channel_id == challenge.channel_id:
+                # !solve called from the challenge - additional args are
+                # assumed to be other solvers.
                 additional_args = args if args else []
-            else:
+            elif channel_id == challenge.ctf_channel_id:
+                # !solve called from the ctf - first argument must be the name
+                # of the challenge.
                 additional_args = args[1:] if len(args) > 1 else []
+            else:
+                # Prevent solves from happening outside the ctf/challenge
+                # rooms.
+                # NOTE: This branch should be unreachable.
+                raise InvalidCommand("You must be in a CTF or Challenge channel to use this command.")
```

Comment:
Let's raise a different exception here, one that doesn't crash, but one that indicates there's a bug.

Reply:
It's already getting caught inside the call to `get_challenge_from_args` and handled using the same "Invalid command" exception, which doesn't crash but returns the message to the user. See attached screenshot for what I mean. This message is generated from line 113 of `util/util.py`.

![error-message](https://user-images.githubusercontent.com/3689038/83936131-6df21e00-a78e-11ea-904e-2b0e085ffabb.png)

If this does end up raising it's because (somehow) the challenge was found but the user is in neither a CTF nor a challenge channel. What should we change the message to say?

Reply:
Resolving this since I rewrote the fix after suggestion from @Kileak